### DISCUSSION TITLE
RMM deployment (ps1 as SYSTEM) - trouble passing arguments

### DISCUSSION THREAD

**Discussion by tmeuze** [Romanitho/Winget-AutoUpdate#500] [Q&A]
Hi there - thank you for the awesome project!

I'm trying to set up ad-hoc deployment using a PowerShell script, to be run in an RMM under SYSTEM context. In order to customize the parameters in `install.bat`, I've hosted a WAU release, with a customized bat file, in my own repo - from which the PowerShell script will pull. 

With what I've conjured up, I can't seem to get any parameters passed to the `Winget-AutoUpdate-Install.ps1`, and I was hoping I could get some advice or help. Maybe someone in the future would find it useful as well.

Here is the PS script which would be executed in the background:
<details>
<summary>deploy.ps1</summary>

```
# install Winget AutoUpdate (WAU)

# set Script Root Directory (srd), Working Directories (wd)
$srd = "c:\APPDIR\WAU"
$wd1 = "$srd\INSTALL"
$wd2 = "$srd\RUNONCE"

# create directories
ForEach ($dir in ("INSTALL", "RUNONCE"))
  {
    New-Item -ItemType "directory" -Force -Path "$srd\$dir"
  }

# download assets
$url1 = "https://github.com/exampleorg/repo/raw/master/WAU.zip"
$url2 = "https://github.com/exampleorg/repo/raw/master/install.bat"
$output1 = "$wd2\WAU.zip"
$output2 = "$wd2\install.bat"
$wc1 = New-Object System.Net.WebClient
$wc1.DownloadFile($url1, $output1)
$wc1.DownloadFile($url2, $output2)

# unzip WAU.zip
Expand-Archive $output1 -DestinationPath $wd1 -Force

# unblock powershell scripts
Get-ChildItem -Path $wd1 -Recurse | Unblock-File

# execute install.bat
Start-Process -FilePath $output2
```

</details>

And the customized batch file:

<details>
<summary>install.bat</summary>

```
SET ScriptParams=-Silent -NotificationLevel None -UpdatesInterval Daily -ListPath https://raw.githubusercontent.com/exampleorg/repo/master/CONFIG/APPS/ -ModsPath https://raw.githubusercontent.com/exampleorg/repo/master/COFIG/MODS/
SET ScriptPath=C:\APPDIR\WAU\INSTALL\Winget-AutoUpdate-Install.ps1

powershell.exe -NoProfile -ExecutionPolicy Bypass -File %ScriptPath% %ScriptParams%
```

</details>

I've tried a few variations of attempting to pass parameters to `Winget-AutoUpdate-Install.ps1` - including merging the execution of it into the aforementioned `deploy.ps1` - but every attempt results in the script executing, but not interpreting the parameters (eg. WAU installer screen appears, does not pull ListPath, etc.). 

I would greatly appreciate any help. Thank you in advance.

**Comment by tmeuze** [comment]
If I disregard `install.bat`, and merge installer script execution into `deploy.ps1`, it still does not interpret the parameters:

<details>
<summary>deploy.ps1</summary>

```
# install Winget AutoUpdate (WAU)

# set Script Root Directory (srd), Working Directories (wd)
$srd = "c:\APPDIR\WAU"
$wd1 = "$srd\INSTALL"
$wd2 = "$srd\RUNONCE"

# create directories
ForEach ($dir in ("INSTALL", "RUNONCE"))
  {
    New-Item -ItemType "directory" -Force -Path "$srd\$dir"
  }

# download assets
$url1 = "https://github.com/exampleorg/repo/raw/master/WAU.zip"
$url2 = "https://github.com/exampleorg/repo/raw/master/install.bat"
$output1 = "$wd2\WAU.zip"
$output2 = "$wd2\install.bat"
$wc1 = New-Object System.Net.WebClient
$wc1.DownloadFile($url1, $output1)
$wc1.DownloadFile($url2, $output2)

# unzip WAU.zip
Expand-Archive $output1 -DestinationPath $wd1 -Force

# unblock powershell scripts
Get-ChildItem -Path $wd1 -Recurse | Unblock-File

$scriptParams = "-Silent -NotificationLevel 'None' -UpdatesInterval 'Daily' -ListPath 'https://raw.githubusercontent.com/exampleorg/repo/master/CONFIG/APPS' -ModsPath 'https://raw.githubusercontent.com/exampleorg/repo/master/COFIG/MODS'"
$scriptPath = "$wd1\Winget-AutoUpdate-Install.ps1"
$cmdString = "$scriptPath $scriptParams"

# execute the command
Invoke-Expression "$cmdString"

```

</details>

Command String prints as:
`c:\APPDIR\WAU\INSTALL\Winget-AutoUpdate-Install.ps1 -Silent -NotificationLevel 'None' -UpdatesInterval 'Daily' -ListPath 'https://raw.githubusercontent.com/exampleorg/repo/master/CONFIG/APPS' -ModsPath 'https://raw.githubusercontent.com/exampleorg/repo/master/COFIG/MODS'`

Which seems right to me?

**Comment by KnifMelti** [comment]
> Command String prints as:
> `c:\APPDIR\WAU\INSTALL\Winget-AutoUpdate-Install.ps1 -Silent -NotificationLevel 'None' -UpdatesInterval 'Daily' -ListPath 'https://raw.githubusercontent.com/exampleorg/repo/master/CONFIG/APPS' -ModsPath 'https://raw.githubusercontent.com/exampleorg/repo/master/COFIG/MODS'`

The command string should be without 'single fnutts' (`-NotificationLevel None` etc.)
Can you really start the `Winget-AutoUpdate-Install.ps1` directly without using `powershell.exe -NoProfile -ExecutionPolicy Bypass -File`

**Reply by tmeuze** [reply]
Thank you for your reply. I've removed the single quotations (to no avail but I think that's a given). 

It seems like installation works just fine when executed directly. In any event, nothing I try seems to work. I've attempted to use variations of:

`& $cmdString`
 `& powershell.exe -NoProfile -ExecutionPolicy Bypass -File $scriptPath $scriptParams`

Resulting in respective errors:
"& : The term $cmdString is not recognized as the name of a cmdlet, function..." 
and
"A parameter cannot be found that matches parameter name 'Silent -NotificationLevel None -UpdatesInterval Daily...'"

Not sure what to try next.

**Reply by KnifMelti** [reply]
I think you need to use `powershell -Command`
Like in https://github.com/Romanitho/Winget-AutoUpdate/blob/main/install.bat

**Reply by tmeuze** [reply]
Thank you. I think I've been trying to over-engineer this, when the included `-Install.ps1` dropped into the RMM will work just fine. 

I appreciate your time!
